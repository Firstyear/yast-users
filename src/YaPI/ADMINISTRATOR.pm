package YaPI::ADMINISTRATOR;

use strict;
use YaST::YCP qw(:LOGGING);
use YaPI;

textdomain ("users");

# ------------------- imported modules
YaST::YCP::Import ("MailAliases");
YaST::YCP::Import ("Users");
# -------------------------------------

our $VERSION		= '1.0.0';
our @CAPABILITIES 	= ('SLES11');
our %TYPEINFO;

=item *
C<$hash Read ();>

Returns the information about system administrator (root).
Currently return hash contains the list of mail aliases.

=cut

BEGIN{$TYPEINFO{Read} = ["function",
    [ "map", "string", "any" ]];
}
sub Read {

    my $self	= shift;

    my $root_mail	= MailAliases->GetRootAlias ();
    return {} if !defined $root_mail;

    my @root_aliases	= ();
    foreach my $alias (split (/,/, $root_mail)) {
	$alias	=~ s/[ \t]//g;
	push @root_aliases, $alias;
    }

    return {
	"aliases"	=> \@root_aliases
    }
}

=item *
C<$string Write ($argument_hash);>

write the system adminstrator data. Supported keys of the argument hash are:

    "aliases"	=> list of mail aliases
    "password"	=> new password
    
Returns error message on error.

=cut

BEGIN{$TYPEINFO{Write} = ["function",
    "string",
    [ "map", "string", "any" ]];
}
sub Write {

    my $self	= shift;
    my $args	= shift;
    my $ret	= "";

    if ($args->{"aliases"} && ref ($args->{"aliases"}) eq "ARRAY") {

	my $root_mail	= join (", ", @{$args->{"aliases"}});
	if (!MailAliases->SetRootAlias ($root_mail)) {
	    # error popup
	    $ret = __("An error occurred while setting forwarding for root's mail.");
	    return $ret;
	}
    }
    if ($args->{"password"}) {
	Users->SetRootPassword ($args->{"password"});
	Users->WriteRootPassword ();
    }
    return $ret;
}

