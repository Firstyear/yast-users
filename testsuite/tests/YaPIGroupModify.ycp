/**
 * $Id$
 */

{
    // testedfiles: Users.pm UsersCache.pm USERS.pm

    include "testsuite.ycp";

    import "YaPI::USERS";
    import "Mode";

    map R = $[
	"passwd": $[
	    "local": $[
		"users": $[
		    500: $[
			"uid"		: "hh",
			"uidnumber"	: 500,
			"type"		: "local",
			"userpassword"	: "x",
			"sn"		: "HH",
			"gidnumber"	: 100,
		    ],
		    "by_name": $[
			"hh":	500,
		    ],
		    "last_uid": 500,
		    "homes":	$[
			"/home/hh":	1,
		    ],
		    "usernames": $[
			"hh":		1,
		    ],
		    "uids": $[
			500:		1,
		    ]
		],
		"shadow": $[
		],
		"groups": $[
		    100: $[
			"cn"	:	"users",
			"gidnumber":	100,
			"type"	:	"local",
			"more_users"	: $[
			    "hh":	1
			],
		    ],
		    "by_name": $[
			"users":	100,
		    ],
		    "gids": $[
			100:		1,
		    ],
		    "groupnames": $[
			"users":	1,
		    ]
		],
	    ],
	    "system": $[
		"users": $[
		    0:	$[
			"uid"		: "root",
			"uidnumber"	: 0,
			"type"		: "system"
		    ],
		    "by_name": $[
			"root"		: 0,
		    ],
		    "last_uid": 0,
		    "homes":	$[
		    ],
		    "usernames": $[
			"root"		: 1,
		    ],
		    "uids": $[
			0		: 1
		    ],
		],
		"shadow": $[
		],
		"groups": $[
		    0: $[
			"cn"		: "root",
			"gidnumber"	: 0,
			"type"		: "system"
		    ],
		    17:	$[
			"cn"		: "audio",
			"gidnumber"	: 17,
			"type"		: "system"
		    ],
		    "by_name": $[
			"root"		: 0,
			"audio"		: 17
		    ],
		    "gids": $[
			0	: 1,
			17	: 1
		    ],
		    "groupnames": $[
			"root"	: 1,
			"audio"	: 1
		    ]
		],
	    ],
	],
	"etc" : $[
	    "fstab": [],
	    "cryptotab": [],
	    "default": $[
		"useradd": $[
		    "home":	"/tmp/local/home",
		    "groups":	"audio,video",
		    "group":	100
		]
	    ],
	],
	"target": $[
	    "stat"	: $[],
	    "size"	: -1,
	    "tmpdir"	: "/tmp/YaST"
	],
	"product": $[
		"features": $[
		    "USE_DESKTOP_SCHEDULER"	: "no",
		    "IO_SCHEDULER"		: "",
		    "ENABLE_AUTOLOGIN"		: "false",
		    "UI_MODE"			: "simple",
		    "EVMS_CONFIG"		: "no"
		]
	],
    ];
    map W = $[];
    map E = $[
	"passwd" : $[
	    "init"	: true
	],
	"target": $[
	    "mkdir"	: true,
	    "bash_output": $[],
	]
    ];

    DUMP ("==========================================================");

    Mode::SetTest ("test");

    map<string,any> config_map  = $[
	"cn"	: "gg",
    ];
    map<string,any> data_map    = $[
	"cn"	: "ggg",
    ];
    E ["target", "bash_output", "stdout"] = "ggg";

    DUMP ("============ edit non-existing group 'gg': =====================");
    TEST (``(YaPI::USERS::GroupModify (config_map, data_map)), [R,W,E], 0);
    DUMP ("============ edit done ========================================");

    config_map    = $[
	"cn"	: "audio",
    ];

    DUMP ("============ edit group 'audio' (change cn): ==================");
    TEST (``(YaPI::USERS::GroupModify (config_map, data_map)), [R,W,E], 0);
    DUMP ("============ edit done ========================================");

    config_map  = $[
	"cn"		: "users",
    ];
    data_map	= $[
	"gidnumber"	: 777
    ];
    E ["target", "bash_output", "stdout"] = "users";

    DUMP ("============ edit group 'users' (change gid): ==================");
    TEST (``(YaPI::USERS::GroupModify (config_map, data_map)), [R,W,E], 0);
    DUMP ("============ edit done ========================================");

    data_map	= $[
	"gidnumber"	: 17
    ];
    E ["target", "bash_output", "stdout"] = "audio";

    DUMP ("============ edit group 'users' (gid conflict): ===============");
    TEST (``(YaPI::USERS::GroupModify (config_map, data_map)), [R,W,E], 0);
    DUMP ("============ edit done ========================================");

}
