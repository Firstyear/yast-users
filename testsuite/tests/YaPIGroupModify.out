Dump	==========================================================
Dump	============ edit non-existing group 'gg': =====================
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.target.size "0" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/usr/lib/yp/yphelper" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/var/lib/YaST2/users.ycp" 0
Read	.target.ycp "/var/lib/YaST2/users.ycp" 0
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $[500:$["gidnumber":100, "sn":"HH", "type":"local", "uid":"hh", "uidnumber":500, "userpassword":"x"], "by_name":$["hh":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_name $["hh":500]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $[100:$["cn":"users", "gidnumber":100, "more_users":$["hh":1], "type":"local"], "by_name":$["users":100], "gids":$[100:1], "groupnames":$["users":1]]
Read	.passwd.local.groups.by_name $["users":100]
Read	.passwd.system.users $[0:$["type":"system", "uid":"root", "uidnumber":0], "by_name":$["root":0], "homes":$[], "last_uid":0, "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_name $["root":0]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $[0:$["cn":"root", "gidnumber":0, "type":"system"], 17:$["cn":"audio", "gidnumber":17, "type":"system"], "by_name":$["audio":17, "root":0], "gids":$[0:1, 17:1], "groupnames":$["audio":1, "root":1]]
Read	.passwd.system.groups.by_name $["audio":17, "root":0]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1]
Read	.passwd.local.groups.groupnames $["users":1]
Read	.passwd.system.groups.gids $[0:1, 17:1]
Read	.passwd.system.groups.groupnames $["audio":1, "root":1]
Return	There is no such group.
Dump	============ edit done ========================================
Dump	============ edit group 'audio' (change cn): ==================
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.target.size "0" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/usr/lib/yp/yphelper" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/var/lib/YaST2/users.ycp" 0
Read	.target.ycp "/var/lib/YaST2/users.ycp" 0
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $[500:$["gidnumber":100, "sn":"HH", "type":"local", "uid":"hh", "uidnumber":500, "userpassword":"x"], "by_name":$["hh":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_name $["hh":500]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $[100:$["cn":"users", "gidnumber":100, "more_users":$["hh":1], "type":"local"], "by_name":$["users":100], "gids":$[100:1], "groupnames":$["users":1]]
Read	.passwd.local.groups.by_name $["users":100]
Read	.passwd.system.users $[0:$["type":"system", "uid":"root", "uidnumber":0], "by_name":$["root":0], "homes":$[], "last_uid":0, "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_name $["root":0]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $[0:$["cn":"root", "gidnumber":0, "type":"system"], 17:$["cn":"audio", "gidnumber":17, "type":"system"], "by_name":$["audio":17, "root":0], "gids":$[0:1, 17:1], "groupnames":$["audio":1, "root":1]]
Read	.passwd.system.groups.by_name $["audio":17, "root":0]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1]
Read	.passwd.local.groups.groupnames $["users":1]
Read	.passwd.system.groups.gids $[0:1, 17:1]
Read	.passwd.system.groups.groupnames $["audio":1, "root":1]
Execute	.target.bash "/bin/cp /etc/group /etc/group.YaST2save" 0
Write	.passwd.groups $["local":$["100":$["cn":"users", "gidnumber":"100", "more_users":$["hh":"1"], "type":"local"], "by_name":$["users":"100"], "gids":$["100":"1"], "groupnames":$["users":"1"]], "system":$["0":$["cn":"root", "gidnumber":"0", "type":"system"], "17":$["cn":"ggg", "gidnumber":"17", "modified":"edited", "more_users":$[], "org_cn":"ggg", "org_gidnumber":"17", "org_group":$["cn":"audio", "gidnumber":"17", "type":"system"], "plugins":[], "removed_userlist":$[], "type":"system", "what":"edit_group"], "by_name":$["audio":"17", "root":"0"], "gids":$["0":"1", "17":"1"], "groupnames":$["audio":"1", "root":"1"]]] true
Execute	.target.bash "/usr/sbin/nscd -i group" 0
Return	
Dump	============ edit done ========================================
Dump	============ edit group 'users' (change gid): ==================
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.target.size "0" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/usr/lib/yp/yphelper" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/var/lib/YaST2/users.ycp" 0
Read	.target.ycp "/var/lib/YaST2/users.ycp" 0
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $[500:$["gidnumber":100, "sn":"HH", "type":"local", "uid":"hh", "uidnumber":500, "userpassword":"x"], "by_name":$["hh":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_name $["hh":500]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $[100:$["cn":"users", "gidnumber":100, "more_users":$["hh":1], "type":"local"], "by_name":$["users":100], "gids":$[100:1], "groupnames":$["users":1]]
Read	.passwd.local.groups.by_name $["users":100]
Read	.passwd.system.users $[0:$["type":"system", "uid":"root", "uidnumber":0], "by_name":$["root":0], "homes":$[], "last_uid":0, "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_name $["root":0]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $[0:$["cn":"root", "gidnumber":0, "type":"system"], 17:$["cn":"audio", "gidnumber":17, "type":"system"], "by_name":$["audio":17, "root":0], "gids":$[0:1, 17:1], "groupnames":$["audio":1, "root":1]]
Read	.passwd.system.groups.by_name $["audio":17, "root":0]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1]
Read	.passwd.local.groups.groupnames $["users":1]
Read	.passwd.system.groups.gids $[0:1, 17:1]
Read	.passwd.system.groups.groupnames $["audio":1, "root":1]
Execute	.target.bash "/bin/cp /etc/group /etc/group.YaST2save" 0
Write	.passwd.groups $["local":$["777":$["cn":"users", "gidnumber":"777", "modified":"edited", "more_users":$["hh":"1"], "org_cn":"users", "org_gidnumber":"777", "org_group":$["cn":"users", "gidnumber":"100", "more_users":$["hh":"1"], "type":"local"], "plugins":[], "removed_userlist":$[], "type":"local", "what":"edit_group"], "by_name":$["users":"100"], "gids":$["100":"1"], "groupnames":$["users":"1"]], "system":$["0":$["cn":"root", "gidnumber":"0", "type":"system"], "17":$["cn":"audio", "gidnumber":"17", "type":"system"], "by_name":$["audio":"17", "root":"0"], "gids":$["0":"1", "17":"1"], "groupnames":$["audio":"1", "root":"1"]]] true
Execute	.target.bash "/bin/cp /etc/passwd /etc/passwd.YaST2save" 0
Write	.passwd.users $["local":$["500":$["gidnumber":"777", "groupname":"users", "modified":"edited", "org_uid":"hh", "org_uidnumber":"500", "sn":"HH", "type":"local", "uid":"hh", "uidnumber":"500", "userpassword":"x", "what":"group_change_default"], "by_name":$["hh":"500"], "homes":$["/home/hh":"1"], "last_uid":"500", "uids":$["500":"1"], "usernames":$["hh":"1"]], "system":$["0":$["type":"system", "uid":"root", "uidnumber":"0"], "by_name":$["root":"0"], "homes":$[], "last_uid":"0", "uids":$["0":"1"], "usernames":$["root":"1"]]] true
Execute	.target.bash "/bin/cp /etc/shadow /etc/shadow.YaST2save" 0
Write	.passwd.shadow $["local":$[], "system":$[]] true
Execute	.target.bash "/usr/sbin/nscd -i passwd" 0
Execute	.target.bash "/usr/sbin/nscd -i group" 0
Return	
Dump	============ edit done ========================================
Dump	============ edit group 'users' (gid conflict): ===============
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.target.size "0" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/usr/lib/yp/yphelper" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/var/lib/YaST2/users.ycp" 0
Read	.target.ycp "/var/lib/YaST2/users.ycp" 0
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $[500:$["gidnumber":100, "sn":"HH", "type":"local", "uid":"hh", "uidnumber":500, "userpassword":"x"], "by_name":$["hh":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_name $["hh":500]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $[100:$["cn":"users", "gidnumber":100, "more_users":$["hh":1], "type":"local"], "by_name":$["users":100], "gids":$[100:1], "groupnames":$["users":1]]
Read	.passwd.local.groups.by_name $["users":100]
Read	.passwd.system.users $[0:$["type":"system", "uid":"root", "uidnumber":0], "by_name":$["root":0], "homes":$[], "last_uid":0, "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_name $["root":0]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $[0:$["cn":"root", "gidnumber":0, "type":"system"], 17:$["cn":"audio", "gidnumber":17, "type":"system"], "by_name":$["audio":17, "root":0], "gids":$[0:1, 17:1], "groupnames":$["audio":1, "root":1]]
Read	.passwd.system.groups.by_name $["audio":17, "root":0]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1]
Read	.passwd.local.groups.groupnames $["users":1]
Read	.passwd.system.groups.gids $[0:1, 17:1]
Read	.passwd.system.groups.groupnames $["audio":1, "root":1]
Return	The group ID entered is already in use.
Dump	============ edit done ========================================
